llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +728,221 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
 }
 
+// Build unlinked shader to ElfPackage with pipeline info.
+//
+// @param [in]  pipelineInfo : Info to build this shader module
+// @param [out] pipelineOut : Output of building this shader module
+// @param [in]  stage : Shader stage of needing to compile
+// @param pipelineDumpFile : Handle of pipeline dump file
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsShaderStage(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                          GraphicsPipelineBuildOut *pipelineOut, UnlinkedShaderStage stage,
+                                          void *pipelineDumpFile) {
+  BinaryData elfBin = {};
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  if (pipelineDumpFile) {
+    std::stringstream strStream;
+    strStream << ";Compiler Options: ";
+    for (auto &option : m_options)
+      strStream << option << " ";
+    std::string extraInfo = strStream.str();
+    PipelineDumper::DumpPipelineExtraInfo(reinterpret_cast<PipelineDumpFile *>(pipelineDumpFile), &extraInfo);
+  }
+
+  Result result = Result::Success;
+  ElfPackage candidateElf;
+
+  // Compile
+  GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+  Context *context = acquireContext();
+  context->attachPipelineContext(&graphicsContext);
+  result = buildUnlinkedShaderInternal(context, shaderInfo, stage, candidateElf, pipelineOut->stageCacheAccesses);
+  releaseContext(context);
+  if (result == Result::Success) {
+    elfBin.codeSize = candidateElf.size();
+    elfBin.pCode = candidateElf.data();
+  }
+
+  if (result == Result::Success) {
+    if (pipelineInfo->pfnOutputAlloc) {
+      void *allocBuf = pipelineInfo->pfnOutputAlloc(pipelineInfo->pInstance, pipelineInfo->pUserData, elfBin.codeSize);
+
+      uint8_t *code = static_cast<uint8_t *>(allocBuf);
+      memcpy(code, elfBin.pCode, elfBin.codeSize);
+
+      pipelineOut->pipelineBin.codeSize = elfBin.codeSize;
+      pipelineOut->pipelineBin.pCode = code;
+    } else {
+      // Allocator is not specified.
+      result = Result::ErrorInvalidPointer;
+    }
+  }
+
+  return result;
+}
+
+// Build the whole graphics pipeline. If missing elfPackage of a certain stage, we will build it firstly, and
+// link them to the full pipeline at last.
+//
+// @param [in]  pipelineInfo    Info to build this shader module
+// @param [out] pipelineOut  : Output of building this shader module
+// @param [in]  elfPackage :  Early compiled elfPackage
```

Comment:
Can you explain that this is an array?

Reply:
> How do you plan to test the new functions?

This change is to prepare for a new feature, some cts tests will be added. We can't  test it now, because there are still some changes in driver.